src/Microsoft.Extensions.Options/IOptionsBuilder.cs

```diff
@@ -0,0 +1,24 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.Extensions.Options
+{
+    /// <summary>
+    /// Represents a builder that configures a named TOptions instance.
+    /// </summary>
+    /// <typeparam name="TOptions">The type of options being requested.</typeparam>
+    public interface IOptionsBuilder<TOptions> where TOptions : class
```

Comment:
Is `IOptionsBuilder` a good name? Because it is meant to forward calls to the `IServiceCollection`. An alternative is to make it agnostic of it, and then have the concrete implementation `flush` the calls to the service collection, but I would think that's a little less intuitive in case you mix using the builder and using Configure directly on the service collection

Reply:
I'd prefer if the only thing you can do with this builder is call Configure as opposed to allowing arbitrary access to IServiceCollection.

i.e. the only operations that would be allowed on the builder would be Configure/PostConfigure (or anything else specifically targeting the builder)

Reply:
Is there any major benefit to having an interface for this as opposed to only the OptionsBuilder class itself, you can still add extension methods in the same way.  There should be nothing that is worth replacing in the base class really.

Reply:
As this is all sugar on top of existing `IServiceCollection` methods

Reply:
There isn't really a benefit for the interface. As it is, it's just syntactic sugar on top of `IServiceCollection`, which is the main challenge I'm addressing. I started this thinking I would store delegates and then have an explicit call that would bind them to the `IServiceCollection`, but that's what I though would be un-intuitive with no real gain (other than decoupling the builder from the DI container temporarily, which IMO is not useful).
So I can change this to be just a public type for us to hook up extension methods.
Still, the specific extension methods would need to access the `IServiceCollection` to be able to forward calls to it, or do you suggest some other approach?

Reply:
Yeah that's fine, we typically hang onto the IServiceCollection in our builders too, and you will definitely have to hang onto the Name as well.  

The main benefit I see is strongly tying a specific named options instance to this builder, so you can just call Configure/PostConfigure without the generic or specifying the name.

I'm still trying to see if we could use this instead of things like `ConfigureApplicationCookie` that we have today.

Reply:
Hehe, configuring the cookie is the only example I saw of using named options in the wild, and I'm not terribly fond of the approach, and hence this proposal. In Orleans we'll have many cases of configuring named services/options.
But even for our singleton service abstractions that can be fulfilled with 1 of different implementations, where each of them might have slightly different typed options, this can be of a lot of use, as the end user will not have to discover which strongly typed options to use, as well as giving the flexibility to chain multiple calls (and not just pick only 1 of the `Action<TOptions>` or `IConfiguration` overloads, in case we provide the extension methods with these overloads).